### PR TITLE
[SPARK-15639][SPARK-16321][SQL] Push down filter at RowGroups level for parquet reader

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

The base class `SpecificParquetRecordReaderBase` used for vectorized parquet reader will try to get pushed-down filters from the given configuration. This pushed-down filters are used for RowGroups-level filtering. However, we don't set up the filters to push down into the configuration. In other words, the filters are not actually pushed down to do RowGroups-level filtering. This patch is to fix this and tries to set up the filters for pushing down to configuration for the reader.

The benchmark that excludes the time of writing Parquet file:

```
test("Benchmark for Parquet") {
  val N = 500 << 12
    withParquetTable((0 until N).map(i => (101, i)), "t") {
      val benchmark = new Benchmark("Parquet reader", N)
      benchmark.addCase("reading Parquet file", 10) { iter =>
        sql("SELECT _1 FROM t where t._1 < 100").collect()
      }
      benchmark.run()
  }
}
```

`withParquetTable` in default will run tests for vectorized reader non-vectorized readers. I only let it run vectorized reader.

When we set the block size of parquet as 1024 to have multiple row groups. The benchmark is:

Before this patch:

The retrieved row groups: 8063

```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_71-b15 on Linux 3.19.0-25-generic
Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
Parquet reader:                          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
reading Parquet file                           825 / 1233          2.5         402.6       1.0X
```

After this patch:

The retrieved row groups: 0

```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_71-b15 on Linux 3.19.0-25-generic
Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
Parquet reader:                          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
reading Parquet file                           306 /  503          6.7         149.6       1.0X
```

Next, I run the benchmark for non-pushdown case using the same benchmark code but with disabled pushdown configuration. This time the parquet block size is default value.

Before this patch:

```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_71-b15 on Linux 3.19.0-25-generic
Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
Parquet reader:                          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
reading Parquet file                           136 /  238         15.0          66.5       1.0X
```

After this patch:

```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_71-b15 on Linux 3.19.0-25-generic
Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
Parquet reader:                          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
reading Parquet file                           124 /  193         16.5          60.7       1.0X
```

For non-pushdown case, from the results, I think this patch doesn't affect normal code path.

I've manually output the `totalRowCount` in `SpecificParquetRecordReaderBase` to see if this patch actually filter the row-groups. When running the above benchmark:

After this patch:
    `totalRowCount = 0`

Before this patch:
    `totalRowCount = 1024000`
## How was this patch tested?

Existing tests should be passed.

### CODE REVIEW

**Comment by viirya** [comment]
@yhuai I am not sure the row-group info is exposed. But I've manually output the `totalRowCount` in `SpecificParquetRecordReaderBase` to check the total number of rows this `RecordReader` will eventually read. The results are shown in the PR description.

**Comment by viirya** [comment]
@liancheng I've updated the benchmark. Please take a look. Thanks.

**Comment by gatorsmile** [comment]
Just realized my PR https://github.com/apache/spark/pull/13728 is related to your PR, especially the description of the two configuration `spark.sql.parquet.filterPushdown` and `spark.sql.parquet.enableVectorizedReader`

**Comment by viirya** [comment]
@gatorsmile yea.

**Comment by gatorsmile** [comment]
Based on my understanding, this performance testing is highly affected by the number of row groups, the number of pruned groups and the number of average rows per group. Do you have any thought about how to control the number of rows per group? Based on the test cases in https://github.com/gatorsmile/spark/blob/9967cc72545324e7a542fcf1b49372d977c0011b/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala#L519-L547, it sounds like each group has only one row... This is strange to me. Do you know how to control it?

**Comment by viirya** [comment]
@gatorsmile No, I have no idea currently. To make sure the pushed down filter is working, I've manually check the `totalRowCount` as shown in PR description.

**Comment by gatorsmile** [comment]
See [the comment](https://github.com/apache/spark/pull/13728#issuecomment-226696567) by @HyukjinKwon the filter is applied more than once. That means, it could be expensive if the filter does not prune a lot of rows. Not sure if my understanding is right.

**Comment by gatorsmile** [comment]
@rdblue Not sure if you can help us? Thank you very much!

**Comment by viirya** [comment]
ping @liancheng @yhuai I've updated the benchmark results. Please see if you have other thoughts. Thanks!

**Comment by viirya** [comment]
@liancheng Is it possible to merge into before 2.0 release?

**Comment by viirya** [comment]
@liancheng @yhuai How about this? Is it ready to merge? Thanks!

**Comment by viirya** [comment]
ping @liancheng @yhuai again...

**Comment by viirya** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 1]
```diff
@@ -85,8 +85,15 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
...
-        l.resolve(fsRelation.dataSchema, fsRelation.sparkSession.sessionState.analyzer.resolver)
+      val dataColumns = l.resolve(fsRelation.dataSchema,
+        fsRelation.sparkSession.sessionState.analyzer.resolver).map { c =>
+          fsRelation.dataSchema.find(_.name == c.name).map { f =>
+            c match {
+              case a: AttributeReference => a.withMetadata(f.metadata)
+              case _ => c
+            }
+          }.getOrElse(c)
+        }
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: Do we need this?

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: I guess a better question is if it is part of the bug fix?

**Comment by yhuai** [comment]
Thank you for the testing. Can you also test the case that a file contains multiple row groups and we can avoid of scanning unneeded ones?

Also since it is not fixing a critical bug, let's not merge it into branch-2.0.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala (hunk 1 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: We use metadata in merged schema to mark the optional field (not existing in all partitions), the metadata is lost after resolving. If we don't add them back, the pushed-down filters will be failed due to non existing field error.

**Comment by viirya** [comment]
@yhuai As I mentioned in the description, I am not sure if we can manipulate row groups as we want, but I have manually tested it to show the actually scanned row numbers.

**Comment by yhuai** [comment]
Sorry. I am not sure I get it. We can set the row group size to a small size. Then, it will not be hard to create a parquet file having multiple row groups.

**Comment by viirya** [comment]
@yhuai Ok. I will have it a try. But looks like I can only manually test it?

**Comment by rdblue** [comment]
@gatorsmile, sorry for the delay, I was evidently not getting notifications until I changed some settings yesterday.

There are a few tests in Parquet that generate files with test data that would be appropriate. The simplest one is [TestReadWriteEncodingStats](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestReadWriteEncodingStats.java#L85-94) where you can see how to build a file with at least one column that has stats you can verify filters with. To get it working for row groups, you'd just need to set the row group size to something small, like 8k, and bump up the number of records until you get a few groups.

Unfortunately, we don't have any pre-built test cases you can use for a performance baseline, though. I think the best you can do with the data generation approach is validate that row groups are being filtered using an approach like you already have, counting the number of rows.

**Comment by gatorsmile** [comment]
@rdblue Really thank you for your reply!

Based on the following code description, https://github.com/Parquet/parquet-mr/blob/master/parquet-column/src/main/java/parquet/filter2/predicate/FilterPredicate.java#L34-L37

It sounds like the invalid rows will still be filtered out during column assembly even if we are unable to prune any row group. If so, counting number of rows does not help us determine how many row groups were pruned. Is my understanding right?

**Comment by rdblue** [comment]
Yeah, Parquet doesn't make a distinction for where filters are applied. If you push a filter, then it will be applied to row groups if possible and individual rows after that. But if you're bypassing the record assembly in Parquet to get vectorized reads, then you don't have to worry about the latter. I'm not sure if Spark does that, since I've not had a chance to get up to speed with Spark's non-Hive read path for Parquet.

What you can do to work around this is to have an ID column that is written in order. Then you can use the column stats to see where the divisions between row groups ends up and set your filters so that row group filtering is the only thing that will be applied.

**Comment by gatorsmile** [comment]
@rdblue Really thank you for your answer! 

@viirya @yhuai I think the enhancement by this PR does not bring the extra penalty that caused by row-level filtering. To confirm it, could @viirya do one more test scenario, as described below? 

Enable filter push down and vectorized read at the same time, but the filters are unable to prune any row group. We can compare the original implementation with the new one. The expected result is no performance difference.

**Comment by viirya** [comment]
I will update this soon..

**Comment by viirya** [comment]
> It sounds like the invalid rows will still be filtered out during column assembly even if we are unable to prune any row group. If so, counting number of rows does not help us determine how many row groups were pruned. Is my understanding right?

@gatorsmile I think it depends. What we do for this filtering is to use `RowGroupFilter.filterRowGroups`. `RowGroupFilter` only performs filtering on row-group level.

**Comment by viirya** [comment]
@yhuai I've run the test to check if row groups are actually filtered. I manually print the number of blocks retrieved in `SpecificParquetRecordReaderBase`. I set the block size to 1024 and rerun the benchmark above. Without this patch, the number of blocks is 8063.  After this patch, the blocks is filtered and the number blocks is 0. So we can make sure that the row groups are actually filtered now.

**Comment by viirya** [comment]
@yhuai BTW, when reading more row groups, the performance improvement is much more.

Before this patch:

Java HotSpot(TM) 64-Bit Server VM 1.8.0_71-b15 on Linux 3.19.0-25-generic
Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
Parquet reader:                          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
reading Parquet file                          1416 / 2168          1.4         691.3       1.0X

After this patch:

    Java HotSpot(TM) 64-Bit Server VM 1.8.0_71-b15 on Linux 3.19.0-25-generic
    Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
    Parquet reader:                          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    reading Parquet file                           246 /  334          8.3         120.3       1.0X

**Comment by gatorsmile** [comment]
@viirya Maybe you have not read my discussion with @rdblue . @rdblue already explained how Parquet internally works. Like what I said above, I think we still need a test for confirming whether there does not exist a noticeable extra penalty when no row is filtered out. Could you do a quick check based on your existing test? Thanks!

**Comment by rdblue** [comment]
@gatorsmile, we've not seen a penalty from running row group level tests when no row groups are filtered and we've decided to turn on dictionary filtering by default. You may see a penalty from using Parquet's internal record-level filter rather than a codegened filter. My recommendation (which is being discussed on the Parquet list) is to add the ability to filter row groups without turning on record-level filters. Should be easy and would solve your problem.

**Comment by gatorsmile** [comment]
@rdblue uh, I see. Thank you for your explanation! My above suggestion is to confirm what you said in @viirya test cases. We expect to see the same results as what you mentioned.

It sounds like dictionary filtering is available in Parquet 1.9. Really look forwarding to it!

**Comment by viirya** [comment]
@gatorsmile This is the benchmark results. No significant difference.

Before this patch:

    Java HotSpot(TM) 64-Bit Server VM 1.8.0_71-b15 on Linux 3.19.0-25-generic
    Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
    Parquet reader:                          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    reading Parquet file                           855 / 1162          2.4         417.2       1.0X

After this patch:

    Java HotSpot(TM) 64-Bit Server VM 1.8.0_71-b15 on Linux 3.19.0-25-generic
    Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
    Parquet reader:                          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    reading Parquet file                           874 / 1228          2.3         426.7       1.0X

**Comment by gatorsmile** [comment]
Great! Around 5% performance penalty looks OK to me. Maybe we can send the code changes to the performance team for seeing TPC-DS improvement. CC @jfchen

**Comment by viirya** [comment]
@gatorsmile You know that the benchmark results will not the same every time even you run it with the same codes. If the difference is under a small range, we can assume they have no significant difference. So I would not say it actually has performance penalty.

**Comment by viirya** [comment]
@gatorsmile  BTW, if the TPC-DS performance is measured with 2.0 codebase, this should benefit the performance.

**Comment by gatorsmile** [comment]
@viirya If you run multiple times and still see 5% performance difference, you can confirm the penalty is around 5%. However, this might also depend on the other factors, e.g., the total time. That means you might see a much lower percentage when the total execution time is very large. 

Hopefully, we can see a performance gain in TPC-DS. That really depends on the workload and the queries. Cross my fingers!

**Comment by viirya** [comment]
@gatorsmile I think the times I run the benchmark is not enough to confirm there is 5% performance difference. But I think it is not important here because we don't want to measure the exact performance difference but only to make sure there is no significant performance penalty.

**Comment by gatorsmile** [comment]
@viirya Yeah, it is not easy to get a full performance picture. I do not know how Spark community did it in the past. When I working for the mainframe team, we had dedicated PQAs for measuring the performance-related topics. We need to present the performance to the customers and give them guides about it. 

How about doing one more test? Changing the way we generate the data set. Insert into the parquet table row by row. Then, you will get many many tiny parquet files. This scenario is not rare in Big Data. I am not sure what is the performance gain and loss in all your test cases.

**Comment by viirya** [comment]
@gatorsmile As I said, it is not an important issue here. What we want to confirm is there is no significant performance penalty..

**Comment by gatorsmile** [comment]
: ) Up to you. I think we will see less perfomance gains and more significant performance penalty when a table contains many many small parquet files. I also think that is expected, anyway.

**Comment by viirya** [comment]
@gatorsmile That depends. In practice, we have many solutions to deal with the case you mentioned. It is not making sense to keep so many tiny parquet files.

**Comment by gatorsmile** [comment]
Agreed. This is the most common performance issue in Big Data. Filter push down will further deteriorate it.

**Comment by viirya** [comment]
ping @liancheng @yhuai Please check if this is ok now. Thanks.

**Comment by viirya** [comment]
ping @liancheng @yhuai again. Can you take a look? Thanks.

**Comment by yhuai** [comment]
@viirya Thank you for updating this. Our schedules are pretty packed for the release. We can take a look at it once 2.0 is released.

**Comment by viirya** [comment]
@yhuai OK. Thanks for letting me know that.

**Comment by viirya** [comment]
@yhuai As 2.0 is released now, do you have time to check this? Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala [hunk 2]
```diff
@@ -357,6 +357,11 @@ private[sql] class ParquetFileFormat
       val hadoopAttemptContext =
         new TaskAttemptContextImpl(broadcastedHadoopConf.value.value, attemptId)
 
+      // Try to push down filters when filter push-down is enabled.
+      // Notice: This push-down is RowGroups level, not individual records.
+      pushed.foreach {
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Without knowing `pushed` is on Option, this make me think that it could be wrong. Since setFilterPredicate() should only be called once, it's better to be like this:
```
if (pushed.isDefined) {
 ParquetInputFormat.setFilterPredicate(hadoopAttemptContext.getConfiguration, pushed.get)
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala (hunk 1 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: @viirya Could you add a regression test for this (it's easy to forget)?

**Comment by davies** [comment]
@viirya In order to have a unit test for this (otherwise it will be broken again in future), we could add some counter in vectorwized parquet reader for row groups for test purpose, then use that to have a regression test.

**Comment by maver1ck** [comment]
I think that this PR also resolves my problem here.
https://issues.apache.org/jira/browse/SPARK-16321?focusedCommentId=15383785&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15383785

**Comment by gatorsmile** [comment]
@maver1ck IMO, in your case, vectorized parquet reader without filter push down performs worse than the original one with filter push down. This is pretty interesting.

**Comment by gatorsmile** [comment]
@maver1ck To use the original reader with filter push down, the best way is to set `spark.sql.parquet.enableVectorizedReader` to false and `spark.sql.parquet.filterPushdown` to true. Let us know if you still see the performance regression. Thanks!

**Comment by maver1ck** [comment]
@gatorsmile 
I added comment in Jira.
"spark.sql.parquet.filterPushdown has true as a default.
Vectorized Reader isn't a case here because I have nested columns (and Vectorized Reader works only with Atomic Types)"

**Comment by gatorsmile** [comment]
uh, I see. This is a regression! We did not call `setFilterPredicate` for the original Parquet reader.

**Comment by gatorsmile** [comment]
@viirya Can you also remove another dead code `overrideMinSplitSize`? 

BTW, it sounds like these removed functions also set multiple internal and external parameters. Should we still set them?

**Comment by gatorsmile** [comment]
Found more out-of-date comments:
- https://github.com/apache/spark/blob/7742d9f1584150befeb2f3d76cdbd4ea1f37c914/sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala#L98
- https://github.com/apache/spark/blob/986b2514013ed9ebab526f2cf3dc714cc9e480bf/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala#L222

Can you update them? Thanks!

**Comment by davies** [comment]
@gatorsmile In order to merge this patch sooner, it's better to only have related changes to fix the regression. We can clean the dead code later.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala (hunk 1 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: @viirya Should we fix that in `l.resolve()` ? cc @marmbrus

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala [hunk 3]
```diff
@@ -581,62 +586,6 @@ private[sql] object ParquetFileFormat extends Logging {
     }
   }
 
-  /** This closure sets various Parquet configurations at both driver side and executor side. */
-  private[parquet] def initializeLocalJobFunc(
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: I looked at all these, seems safe to remove them (also overrideMinSplitSize).

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: @davies Do you mind if I ask a question please? So, does removing overrideMinSplitSize mean  https://issues.apache.org/jira/browse/SPARK-10143 is not happening anymore? I thought this is a regression that we don't use this function.

I am pretty sure that you took a look for this as well but just want to double check and for my curiosity.

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: I discussed this with @yhuai offline, because the default value of split size (128M) could match the default value of group size (128M) in parquet, so we will not regression in most the cases. It's true that user may still have a different value in HadoopConf, we did not respect that anymore, I think it's fine.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Thanks for kind explanation!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 4]
```diff
@@ -527,4 +536,43 @@ class ParquetFilterSuite extends QueryTest with ParquetTest with SharedSQLContex
       assert(df.filter("_1 IS NOT NULL").count() === 4)
     }
   }
+
+  test("Fiters should be pushed down for vectorized Parquet reader at row group level") {
```
Comment (maver1ck) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: What about non-vectorized reader ?

Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: For non-vectorized reader, we use parquet's `ParquetRecordReader` and push the filters into.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 5]
```diff
@@ -527,4 +536,43 @@ class ParquetFilterSuite extends QueryTest with ParquetTest with SharedSQLContex
...
+      val path = s"${dir.getCanonicalPath}/table"
+      (1 to 1024).map(i => (101, i)).toDF("a", "b").write.parquet(path)
+
+      val requiredSchema = StructType(Seq(
+        StructField("a", IntegerType, nullable = false),
+        StructField("b", IntegerType, nullable = false)
+      ))
+
+      val hadoopConf = new Configuration()
+      hadoopConf.setInt("parquet.block.size", 1)
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: Do you mind if I ask why preferred row group size should be 1 byte?

Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: Just to make sure it produces many row groups.

Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: Ah, I see. Thanks. Maybe this should be commented.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala (hunk 1 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: When I add a regression test, I find that this change is not needed. Not sure which recent commit helps keeping metadata after resolving. I will add the regression test still.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala (hunk 3 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Looks safe to remove overrideMinSplitSize. So I also delete it in new commit.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala (hunk 4 above)
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: We already have test for the pushed down filters for ParquetRecordReader. But it is at individual record level. If you mean row group level, because ParquetRecordReader doesn't expose a count for the row group, I think we can't check if the filter is pushed down. Besides, it seems to be the functionality of ParquetRecordReader, and I think it should be unit test in parquet project.

Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: Yea, as the ParquetRecordReader also uses the Configuration to get the pushed down filters, I think this also fixes SPARK-16321.

**Comment by gatorsmile** [comment]
Could you also change the PR title and add SPARK-16321?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 6]
```diff
@@ -527,4 +538,54 @@ class ParquetFilterSuite extends QueryTest with ParquetTest with SharedSQLContex
...
+      val filters = ParquetFilters.createFilter(requiredSchema, sources.LessThan("a", 100))
+      assert(filters.isDefined)
+
+      val attemptId = new TaskAttemptID(new TaskID(new JobID(), TaskType.MAP, 0), 0)
+      val hadoopAttemptContext1 =
+        new TaskAttemptContextImpl(hadoopConf, attemptId)
+      val hadoopAttemptContext2 =
+        new TaskAttemptContextImpl(hadoopConf, attemptId)
+
+      ParquetInputFormat.setFilterPredicate(hadoopAttemptContext1.getConfiguration, filters.get)
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: I think we should test the ParquetInputFormat to do the push down, otherwise we are testing parquet itself.

In order to known how many row groups are read, we could have a SQL metrics for that.

Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: ok. Let me update it. Thanks.

**Comment by viirya** [comment]
@davies I improved the test case by adding a SQL metrics. Please take a look. Thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 7]
```diff
@@ -368,73 +378,75 @@ class ParquetFilterSuite extends QueryTest with ParquetTest with SharedSQLContex
...
-
-        val df2 = spark.read.parquet(pathSeven, pathEight).filter("a = 1").selectExpr("a", "b")
-        checkAnswer(
-          df2,
-          Row(1, "1"))
-
-        // The fields "a" and "b" exist in both two Parquet files. No metadata is set.
-        assert(!df2.schema("a").metadata.contains(StructType.metadataKeyForOptionalField))
-        assert(!df2.schema("b").metadata.contains(StructType.metadataKeyForOptionalField))
+    Seq("true", "false").map { vectorized =>
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: Is this change related?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 8]
```diff
@@ -199,6 +209,19 @@ private[sql] case class FileSourceScanExec(
         options = relation.options,
         hadoopConf = relation.sparkSession.sessionState.newHadoopConfWithOptions(relation.options))
 
+      (file: PartitionedFile) => {
+        val iter = func(file)
+        // Only for test purpose.
+        // Once the vectorized Parquet reader is initialized in the above method, we can read its
+        // variable numRowGroups.
+        if (fileFormat != null) {
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: These looks ugly, can we create a named accumulator here, then pass that into VectorizedParquetRecordReader and update it there?

We could add an private api in AccumulatorContext to lookup by name.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala (hunk 7 above)
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: A regression test added for the optional column (only existing in part of parquet files). Previously it is only for non vectorized parquet reader. Now adding test for vectorized reader. This PR has related changes before but it is removed as in the discussion folded.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 8 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: hmm, VectorizedParquetRecordReader is not exposed to outside of ParquetFileFormat. It is wrapped in a returned anonymous function that takes a PartitionedFile and read data from it. In order to pass it to VectorizedParquetRecordReader, we might need to change current API of FileFormat. Is it worth doing that for this?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: I think we can directly update the accumulator in ParquetFileFormat if there is the accumulator. I will update the codes later and please check if it is good for you.

**Comment by viirya** [comment]
ping @davies Please review the latest changes. Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 9]
```diff
@@ -213,7 +213,9 @@ private[sql] case class FileSourceScanExec(
 
   private[sql] override lazy val metrics =
     Map("numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"),
-      "scanTime" -> SQLMetrics.createTimingMetric(sparkContext, "scan time"))
+      "scanTime" -> SQLMetrics.createTimingMetric(sparkContext, "scan time"),
+      // Only for test purpose.
+      "numRowGroups" -> SQLMetrics.createMetric(sparkContext, "numRowGroups"))
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Can we only create this for unit test (manually create in test case)?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala [hunk 10]
```diff
@@ -357,10 +358,27 @@ private[sql] class ParquetFileFormat
...
         logDebug(s"Appending $partitionSchema ${file.partitionValues}")
+
+        // For test purpose.
+        // If the predefined accumulator exists, the row group number to read will be updated
+        // to the accumulator. So we can check if the row groups are filtered or not in test case.
+        val taskContext = TaskContext.get()
+        if (taskContext != null) {
+          val accu = taskContext.taskMetrics.lookForAccumulatorByName(accuNameForNumRowGroup)
+          if (accu.isDefined) {
+            accu.get.asInstanceOf[SQLMetric].add(vectorizedReader.getRowGroupCount().toLong)
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: We could move this into SpecificParquetRecordReaderBase to minimize the change.

**Comment by davies** [comment]
LGTM, could you fix the conflict (should be trivial)?

**Comment by viirya** [comment]
@davies Thanks. I've fix it. Waiting for jenkins tests.

**Comment by viirya** [comment]
@davies The accumulator could be released by JVM early for optimization. I made new change to prevent it.

**Comment by viirya** [comment]
an unrelated test failed...

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
The use of accumulator is unstable...

**Comment by viirya** [comment]
@davies Due to the accumulators for vectorized setting "true" and "false" are with the same names "numRowGroups", the API looking for the accumulator will return one among them. So the test becomes unstable. By removing it from `AccumulatorContext` at the end of test, it should solve this problem.

**Comment by davies** [comment]
Merging this into master and 2.0, thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 11]
```diff
@@ -527,4 +530,32 @@ class ParquetFilterSuite extends QueryTest with ParquetTest with SharedSQLContex
...
+        val path = s"${dir.getCanonicalPath}/table"
+        (1 to 1024).map(i => (101, i)).toDF("a", "b").write.parquet(path)
+
+        Seq(("true", (x: Long) => x == 0), ("false", (x: Long) => x > 0)).map { case (push, func) =>
+          withSQLConf(SQLConf.PARQUET_FILTER_PUSHDOWN_ENABLED.key -> push) {
+            val accu = new LongAccumulator
+            accu.register(sparkContext, Some("numRowGroups"))
+
+            val df = spark.read.parquet(path).filter("a < 100")
+            df.foreachPartition(_.foreach(v => accu.add(0)))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: what does this test? shouldn't `accu` always be zero?

Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: In `SpecificParquetRecordReaderBase`, it looks for an accumulator `numRowGroups` if any and update it with the row group number to read. It is for test purpose only.

Here we force this trivial foreach function refer this accumulator, but doesn't change it.
